optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/test/java/org/optaplanner/quarkus/constraints/OptaPlannerProcessorConstraintProviderAndDrlTest.java

```diff
@@ -1,70 +0,0 @@
-/*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package org.optaplanner.quarkus.constraints;
-
-import static org.junit.jupiter.api.Assertions.*;
-
-import java.util.Collections;
-
-import javax.inject.Inject;
-
-import org.jboss.shrinkwrap.api.ShrinkWrap;
-import org.jboss.shrinkwrap.api.spec.JavaArchive;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.RegisterExtension;
-import org.optaplanner.core.api.solver.SolverFactory;
-import org.optaplanner.core.config.solver.SolverConfig;
-import org.optaplanner.quarkus.KieRuntimeBuilderMock;
-import org.optaplanner.quarkus.deployment.config.OptaPlannerBuildTimeConfig;
-import org.optaplanner.quarkus.testdata.normal.constraints.TestdataQuarkusConstraintProvider;
-import org.optaplanner.quarkus.testdata.normal.domain.TestdataQuarkusEntity;
-import org.optaplanner.quarkus.testdata.normal.domain.TestdataQuarkusSolution;
-
-import io.quarkus.deployment.builditem.CapabilityBuildItem;
-import io.quarkus.test.QuarkusUnitTest;
-
-public class OptaPlannerProcessorConstraintProviderAndDrlTest {
```

Comment:
@Christopher-Chianelli This is a side-effect of my change. Specifically:

    The kieBaseSupplier (org.optaplanner.quarkus.gizmo.OptaPlannerDroolsInitializer$Implementation$$function$$1@ba17be6) is not a KieBaseDescriptor. Maybe remove calls to ScoreDirectorFactoryConfig.setKieBaseSupplier(Supplier)?
    	at org.optaplanner.core.impl.score.director.stream.DroolsConstraintStreamScoreDirectorFactory.assertIsKieBaseDescriptor(DroolsConstraintStreamScoreDirectorFactory.java:83)
	    at org.optaplanner.core.impl.score.director.stream.DroolsConstraintStreamScoreDirectorFactory.<init>(DroolsConstraintStreamScoreDirectorFactory.java:89)
	    
I think this happens because:
- `SolverFactory` now creates `ScoreDirectorFactory` in the constructor,
- and therefore if the constructor gets a wrong `SolverConfig`, it throws this exception immediately.

The situation that you were testing therefore can not happen, this exception will be thrown earlier.
That said, I am not too excited about deleting this test - can you suggest a better approach, with your knowledge of the OP-Quarkus codebase?

Reply:
This happens because DRL need to get the generated KieBase in Quarkus, so a custom Supplier is written and passed to GizmoKieBaseSupplier (which get said generated KieBase). However, the KieBase Supplier required by DroolsConstraintStream is required to be a particular type, and since the same KieBase Supplier is used for both Drools score calculation method, this exception is thrown.

The situation being tested is a helpful error message is thrown if both DRL and Constraint Streams are used (in particular, both DRL on classpath + ConstraintProvider.

We might be able to throw this exception at Quarkus build time. What happens if you set a constraint provider + score drl in base OptaPlanner now?

Reply:
Look like this happens because all the factories are built first, and then the "only one" factory constraint is validated.

Reply:
The fix would be to either validate the "only one" score directory factory constraint before the factories are built, or to throw the exception at Quarkus build time (which will do the "only one" score directory factory constraint check).

Reply:
I was thinking the same. I may have to do the SPI in two stages - first check if configured, then instantiate if selected. That way, I will be able to avoid touchich Quarkus altogether.